Rock.Messaging.RabbitMQ/RabbitEngine.cs

```diff
@@ -0,0 +1,105 @@
+        }
+
+        public void Dispose()
+        {
+            _connection.Dispose();
+            _consumerModel.Abort();
+        }
+
+        #region Receiver
+        public void Start(string selector) // TODO: Discuss: RabbitMQ supports getting individual messages- should Rock.Messaging support that?
+        {
+            _consumerModel = _connection.CreateModel();
+            _consumerModel.ExchangeDeclare(_config.Exchange, _config.ExchangeType ?? (topical ? "topic": "direct"), true); // Should perhaps configure this default somehow? Also RabbitMQ Topics might not fulfill the same intent as the abstraction?
+            _consumerModel.QueueDeclare(_config.QueueName, true, false, false, null); // TODO: Discuss: Add config settings for these? It's already pretty cluttered...
+            if (!string.IsNullOrWhiteSpace(selector))
+            {
+                _consumerModel.QueueBind(_config.QueueName, _config.Exchange, selector);
+            }
+            else
+            {
+                _consumerModel.QueueBind(_config.QueueName, _config.Exchange, _config.RoutingKey ?? string.Empty);
+            }
+            
+            var consumer = new EventingBasicConsumer(_consumerModel);
+            consumer.Received += delegate(object sender, BasicDeliverEventArgs args)
+            {
+                var messageReceived = new MessageReceivedEventArgs(new RabbitReceiverMessage(args, _consumerModel));
+                OnMessageReceived(messageReceived);
+                if (_config.AutoAcknowledge)
+                {
+                    _consumerModel.BasicAck(args.DeliveryTag, false);
+                }
+            };
+            _consumerModel.BasicQos(0, _config.MaxRequests, true);
+            _consumerModel.BasicConsume(_config.QueueName, false, consumer);
+        }
+
+        protected virtual void OnMessageReceived(MessageReceivedEventArgs args)
+        {
+            var handler = MessageReceived;
+
+            if (handler != null)
+            {
+                handler(this, args);
+            }
+        }
+        public event EventHandler<MessageReceivedEventArgs> MessageReceived;
+
+        #endregion
+        public Task SendAsync(ISenderMessage message)
+        {
+            return Task.Run(delegate
+            {
+                using (var model = _connection.CreateModel())
+                {
+                    var props = model.CreateBasicProperties();
+                    props.Headers = message.Headers.ToDictionary<KeyValuePair<string, string>, string, object>(kvp => kvp.Key, kvp => kvp.Value); // IEnum<KVP<S,S>> in, Dictionary<string, object> out.
+                    props.ContentType = message.MessageFormat.ToString();
+                    props.Priority = message.Priority ?? 0;
+                    // props.ContentEncoding = new NotImplementedException(); // TODO: add this to Rock.Messaging?
+                    model.BasicPublish(_config.Exchange, _config.RoutingKey, props, message.BinaryValue);
```

Comment:
@mjeaton:
> Publishing with a RoutingKey seems odd, especially from the config. What if we want this per message? Publishing with a RoutingKey this way is very different from how other systems do it.

@dtarjeft:
> The documentation got a little sparse around here, I made this call because there are 3 overloads for "BasicPublish" and they appear to all involve a RoutingKey at some point or another... I believe one can pass in null for that (i.e. leave it out of the config) and have a "default" binding on the exchange, but that's getting into design. :)

> Worth mentioning about RabbitMQ: Publishers have no sense of "queue", they only have sense of "exchange"- so RoutingKey is effectively equivalent to the Publisher having a Queue Name in other systems. Publishers say "Hey, this is what I'd like for you to do with my message" and hopefully Exchanges have that key bound to a queue from a receiver.

We need to have a discussion about this. In other messaging systems, when a receiver listens with a "selector", the selector is used to compare against a message's properties and potentially their values. RabbitMQ seems to be significantly different here.